Plex media player component

miniconfig opened the pull request:
Currently supports metadata and play/pause of video.

miniconfig committed:
Initial version of plex media player component

miniconfig committed:
Added requirements

miniconfig committed:
Fixed Requirements URL

miniconfig committed:
Fixed additional pylint and flake issues

luxus commented:
don't know how to hand over this small patch 
https://github.com/luxus/home-assistant/commit/7ee97e648ef625eef4b6e0e741708c0d841a4cd0

SEJeff commented on homeassistant/components/media_player/plex.py:
```diff
@@ -0,0 +1,199 @@
+            elif state == 'paused':
+                return STATE_PAUSED
+            else:
+                return STATE_UNKNOWN
```
This is really more of a stylistic suggestion than anything, but please always leave a default "fall through return" for things like this. It makes for a lot less bugs in the future when more stuff might be added to this or this code might get refactored. ie:

```python
def state(self):
        if self._media is None:
            return STATE_IDLE
        else:
            state = self._media.get('state')
            if state == 'playing':
                return STATE_PLAYING
            elif state == 'paused':
                return STATE_PAUSED
    return STATE_UNKNOWN
```

SEJeff commented on homeassistant/components/media_player/plex.py:
```diff
@@ -0,0 +1,199 @@
+        """ Content type of current playing media. """
+        if self._media is None:
+            return None
+        else:
```
In the same vein as the previous comment, the else in this context can not execute if the return does, so it isn't strictly necessary. You can remove the else and left indent the rest of this property.

SEJeff commented on homeassistant/components/media_player/plex.py:
```diff
@@ -0,0 +1,199 @@
+        if self._media is not None:
+            return self.server.library.getByKey(self.media_content_id).thumbUrl
+        else:
+            return None
```
For this and several of the below methods that follow this pattern, would you mind removing the else and left indenting the `return None` please?

balloob commented on homeassistant/components/media_player/plex.py:
```diff
@@ -0,0 +1,199 @@
+
+REQUIREMENTS = ['https://github.com/miniconfig/python-plexapi/archive/'
+                '437e36dca3b7780dc0cb73941d662302c0cd2fa9.zip'
+                '#python-plexapi==1.0.2.5']
```
This version needs to match the exact version that is going to be installed. In this case that would be 1.0.2. Otherwise it will keep installing it.

miniconfig replied:
Ah, sorry about that.  I changed the version to 1.0.2.5 because I was concerned about my fork being confused with the "official" python-plexapi which is at 1.0.2 (and I figured the next official version could very well be 1.0.3) - Should I change my fork to be 1.0.2.5, or is my concern unfounded in the first place?

balloob replied:
Python versioning does not allow 4 numbers. Keep it at 1.0.2 or otherwise once 1.0.3 comes out we have no way of distinguishing it.

snizzleorg commented:
is this already in the main branch?

SEJeff commented:
@snizzleorg nope but with those comments addressed this will make it to the dev branch

miniconfig committed:
Fixed various property methods to make sure they all had a fall through return and removed unnecessary "else" statements

miniconfig committed:
Removed references to the frontend device parameter in the directions and added some clarification.
Fixed plexapi version number.

miniconfig committed:
Pylint errors

SEJeff commented on homeassistant/components/media_player/plex.py:
```diff
@@ -0,0 +1,188 @@
+    SUPPORT_NEXT_TRACK, MEDIA_TYPE_TVSHOW, MEDIA_TYPE_VIDEO)
+from homeassistant.const import (
+    STATE_IDLE, STATE_PLAYING, STATE_PAUSED, STATE_UNKNOWN)
+from plexapi.myplex import MyPlexUser
```
One final thing, which will fix the build. To make this an optional platform, and hence `plexapi` dependency, can you delay the import? This is generally a python anti-pattern, but is done in this project to allow shipping all of the code together, but only installing the requirements, when the dependencies are installed.

Move:

```python
from plexapi.myplex import MyPlexUser
```

into `setup_platform()`. That will also have the benefit of fixing the build.

The changes look great! Thankyou for humoring me here.

miniconfig committed:
Added plexapi library to requirements_all.txt

balloob commented:
Looks good. We can merge after @SEJeff issue has been addressed.

miniconfig committed:
Moved plexapi import into setup_platform().
Changed CONTRIBUTING.md to refer to requirements_all.txt instead of requirements.txt

SEJeff commented:
Thanks!
